Finished first version of the task2

ArteIIo opened the pull request:
Task two for review.

ArteIIo committed:
Finished first version of the task1

Added CRUD realisation and played with some features.

ArteIIo committed:
Minore Fixes

Add minore fixes after review.

ArteIIo committed:
Done task 2

Done task two. Created a version for review.

ArteIIo committed:
Done second part

Added some features, corrected some moments.

ArteIIo committed:
Complited task for review

Change some moments (mostly DataProvider class).

anton-borodulin-softserve commented on BookService/BookService/Controllers/AuthorsController.cs:
```diff
@@ -0,0 +1,143 @@
+        /// </summary>
+        /// <param name="id">Index of the needed author</param>
+        /// <returns>Ok if there is a author by such index</returns>
+        // GET api/ authors/5
```
this is default comments? In any way, if you want any additional message put it after line 47.  The summary block can be multiline.

anton-borodulin-softserve commented on BookService/BookService/Controllers/AuthorsController.cs:
```diff
@@ -0,0 +1,143 @@
+namespace BookService.Controllers
+{
+    /// <summary>
+    /// Controller for CRUD opertions
```
There is a nice style to put a dot. At the end of any comment line.

anton-borodulin-softserve commented on BookService/BookService/Controllers/AuthorsController.cs:
```diff
@@ -0,0 +1,143 @@
+        /// <param name="id">Index of the selected book</param>
+        /// <param name="author">New auhtor's values</param>
+        /// <returns>CreateAtAction result if author
+        /// has been updated or bad request otherwise</returns>
```
Same 94 and 95 line should be moved up after 89

anton-borodulin-softserve commented on BookService/BookService/Controllers/AuthorsController.cs:
```diff
@@ -0,0 +1,143 @@
+        /// <returns>Ok if there is a author by such index</returns>
+        // GET api/ authors/5
+        [HttpGet("{id}")]
+        public IActionResult Get(int id)
```
Don't use names for methods like Get or Update etc. It's already can be viewed by attributes. Use names like GetBook, UpdateAuthor.

anton-borodulin-softserve commented on BookService/BookService/Controllers/AuthorsController.cs:
```diff
@@ -0,0 +1,143 @@
+        /// has been created or bad request otherwise</returns>
+        // POST api/ Authors
+        [HttpPost]
+        public IActionResult Post([FromBody] Author author)
```
+1

anton-borodulin-softserve commented on BookService/BookService/Controllers/AuthorsController.cs:
```diff
@@ -0,0 +1,143 @@
+        /// has been updated or bad request otherwise</returns>
+        // PUT api/books/5
+        [HttpPut("{id}")]
+        public IActionResult Put(int id, [FromBody] Author author)
```
+1

anton-borodulin-softserve replied:
Do you use id zero-based or 1? if 1, 0 id is incorrect and should be validated. I proposed option with nullable id in some request model with a required attribute but also can be checked with data annotation min value or if etc.  this is not mandatory to fix, just keep in mind.

anton-borodulin-softserve commented on BookService/BookService/Controllers/BooksController.cs:
```diff
@@ -0,0 +1,188 @@
+        /// <summary>
+        /// Remove book's author
+        /// </summary>
+        /// <param name=id">Index of the book</param>
```
Here is missing ". ="id".

anton-borodulin-softserve commented on BookService/BookService/Startup.cs:
```diff
@@ -0,0 +1,49 @@
+            ContainerBuilder builder = new ContainerBuilder();
+            builder.Populate(services);
+            builder.RegisterType<DataProvider>().As<IDataProvider>();
+            builder.RegisterType<LibraryCollection>().As<ILibrary>().SingleInstance();
```
Do you still need SingleInstance for LibraryCollection ?

anton-borodulin-softserve commented on BookService/Logic/DataProvider.cs:
```diff
@@ -0,0 +1,59 @@
+
+            books = new List<Book>()
+            {
+                new Book() { Id = 0, Name = "Book0", Author = authors[0] },
```
starting from id 0 is bad style as 0 is default(int)

anton-borodulin-softserve commented on BookService/Logic/ILibrary.cs:
```diff
@@ -0,0 +1,113 @@
+        /// <returns>Book by selected index</returns>
+        /// <exception cref="ArgumentOutOfRangeException">Throw when index out of
+        /// list's count range</exception>
+        Book GetBookByIndex(int id);
```
why the parameter is not calling index if it is an index?  At least the documentation says in this way.

anton-borodulin-softserve commented on BookService/Logic/LibraryCollection.cs:
```diff
@@ -0,0 +1,172 @@
+        /// <param name="id">Index of the new book</param>
+        /// /// <exception cref="IndexOutOfRangeException">Throw when index out of
+        /// list's count range</exception>
+        public void SetBookByIndex(Book book, int id)
```
id should be changed to index.

anton-borodulin-softserve commented on BookService/Logic/LibraryCollection.cs:
```diff
@@ -0,0 +1,172 @@
+        public Book RemoveBook(int id)
+        {
+            Book deleted = books[id];
+            books.RemoveAt(id);
```
I understand. You had messed up id and index. The Identifier shouldn't be related to the index. If you want to optimize select`s based on index. You should used Find() it`s fast. If you want to optimize even more. You can store your collection in dictionary<int, book> in this way id will be duplicated, one as key for the dictionary and one as the property of book. Also disadvantages that you sill need to manage that. But it will work fast and in the same way as you are doing here. But I advise you to use Find() as if you will choose dictionary you will need to delete this logic when you will work at 4 level. But in some other case, it can be useful.

anton-borodulin-softserve commented on BookService/Tests/AuthorTests.cs:
```diff
@@ -0,0 +1,154 @@
+        public void Library_GetAuthorByIndex_Exception(int index)
+        {
+            // Arrage
+            InitLists();
```
You know in case if you don't assign a great amount of data for the test. Initialize data in the test, not in a separate method. If it will lead not a big amount of duplicated code! 
 Because in this case, it will be easier to maintain and read. Now I need for example to read method name then find InitList and what it`s added to the collection come back here again. etc.

anton-borodulin-softserve commented on BookService/Tests/AuthorTests.cs:
```diff
@@ -0,0 +1,154 @@
+        /// Test for SetAuthorByIndex-method
+        /// </summary>
+        [Fact]
+        public void Library_SetAuthorByIndex()
```
You can use a name like Library_SetAuthorByIndex_Positive or Library_SetAuthorByIndex_Correct_Result or to mark in the end what behavior you are testing.

anton-borodulin-softserve commented on BookService/Tests/TestsBase.cs:
```diff
@@ -0,0 +1,53 @@
+        {
+            Mock<IDataProvider> data = new Mock<IDataProvider>();
+
+            data.Setup(p => p.GetAuthors()).Returns(authors);
```
Here you can use one line assignment. like  data.Setup(p => p.GetAuthors()).Returns(new List<Author>{new Author() { Id = 0, Name = "Name0", Surname = "Surname0" }});  and use this like in each test if it is needed there. But if you need not 1-5 line of assigment you shoud to consider create some method that will be reused.

anton-borodulin-softserve reviewed (approved):
Index == Id is bad idea. You should change that.
 All other comments are more like cosmetical. Looks good.

ArteIIo committed:
Fixes after review

Fixed main problems (First of all it was ID==index), and made some cosmetic moments.
